src/Jackalope/Version/VersionHandler.php

```diff
@@ -119,27 +129,63 @@ public function checkinItem($path)
 
         // TODO set subgraph to read only
 
-        $versionHistoryNode = $node->getPropertyValue('jcr:versionHistory');
+        /** @var NodeInterface $versionHistoryNode */
+        $versionHistoryUuid = $node->getProperty('jcr:versionHistory')->getString();
+        $versionHistoryNode = $this->objectManager->getNodeByIdentifier($versionHistoryUuid, 'Version\VersionHistory');
 
         // FIXME add some kind of sharding
         // should avoid to have too many nodes on the same level
+        /** @var NodeInterface $versionNode */
         $versionNode = $versionHistoryNode->addNode(UUIDHelper::generateUUID(), 'nt:version');
         $versionNode->setProperty('jcr:uuid', UUIDHelper::generateUUID());
         $versionNode->setProperty('jcr:created', new \DateTime());
+        $versionNode->setProperty('jcr:successors', array());
+
+        $this->createFrozenNode($versionNode, $node);
 
-        // TODO create frozen node
+        $baseVersionNode->setProperty('jcr:successors', array($versionNode), PropertyType::REFERENCE, false);
+        $baseVersionNode->setModified();
+        $versionNode->setProperty(
+            'jcr:predecessors',
+            $node->getProperty('jcr:predecessors')->getString(),
+            PropertyType::REFERENCE
+        );
 
-        $versionNode->setProperty('jcr:predecessors', $node->getProperty('jcr:predecessors')->getString());
-        $node->setProperty('jcr:predecessors', array(), PropertyType::REFERENCE, false);
-        foreach ($versionNode->getProperty('jcr:predecessors') as $predecessorUuid) {
-            $predecessor = $this->objectManager->getNodeByIdentifier($predecessorUuid);
-            $predecessor->setProperty('jcr:successors', array_merge($predecessor->getPropertyValueWithDefault('jcr:successors', array()), array($versionNode)));
+        // FIXME Find a better way to solve the referencing not existing nodes issue
+        $this->session->save();
```

Comment:
Tried to remove it, but it didn't work, although I have set a UUID on the node to be referenced...